### DISCUSSION TITLE
Interpretation of parameters under non-identity link functions

### DISCUSSION THREAD

**Discussion by eort** [lnccbrown/HSSM#394] [Q&A]
Hi, 

Thanks for fixing the issues regarding sampling when the prior is not (-inf, inf). Using `link_settings='log_logit'` improves my fits massively!

I was wondering about the interpretation though. From replies in discussion (primarily by @AlexanderFengler), I understand, that under these link function settings, the parameters cannot be interpreted directly, but need to be transformed back to the "standard space" in order for its values to make sense in the original sense.  Is it correct that this back-transformation can be done by calling `model.link['<parameter>'].linkinv(x)`? If so, can this be done on all samples of the `inference data` object in the model after sampling, such that calling the arviz plotting function would automatically take the back-transformed values rather that the fitted ones?

Perhaps to give a concrete example. I fitted following model:

```
model = hssm.HSSM(data=data,
                  model='angle',
                  include=[
                       {
                        "name": "a",
                        "formula":"a ~ C(drug, levels=lvl) + (1 | participant_id)", # drug levels are "plac" and "cyc"
                        "prior": {
                            "Intercept": {"name": "Normal", "mu": 0.5, "sigma": 0.5},
                            "C(drug, levels=lvl)": {"name": "Normal", "mu": 0, "sigma": 1}
                        }],
                  extra_namespace={"lvl": cfg['drugs']},
                  hierarchical=True,
                  link_settings='log_logit')
```
resulting in following fit:
![image](https://github.com/lnccbrown/HSSM/assets/8819465/3ca7f6fe-24cb-447d-a125-f90e26a7e541)

Ignoring the random effects, is it correct to conclude that posterior of `a` peaks around 0.97 (`linkinv(-1.1)`) for  the intercept (i.e. condition "plac"), ranging from  0.83 (`linkinv(-1.1)`) to  1.13 (`linkinv(-0.8)`), whereas the influence of the condition "cyc", relative to the intercept, is described by a posterior from 1.68(`linkinv(0.05)`) to 1.72 (`linkinv(0.11)`) with the mode being at 1.70 (`linkinv(0.08)`?

Does this make sense?

Thanks!

**Comment by digicosmos86** [comment]
Hi @eort,

I'll defer to @AlexanderFengler on this if he disagrees, but I think your intuition is correct. In general the intuition of link functions in GLMs such as logistic regressions also applies here, so the interpretation of the parameter estimates should be made after transforming them back to their original space using the inverse functions.

**Comment by AlexanderFengler** [comment]
Hi @eort ,

you are right, basically apply the correct inverse link for each parameter.
We are working on including a bit of convenience functionality on that front, hopefully by end of next week there should be a PR on that. 

Thank you for trying HSSM :).

Best,
Alex

**Comment by eort** [comment]
Hi both of you,

Thanks for your answers! 

If you don't mind I have a follow-up. I realized that after the back-transformation, my estimates are a little unrealistic. Specifically, while the intercept in my regression (`a ~ drug + ( 1 | participant_id)`) is in a reasonable range, the drug effect is extremely large. See below for the back-transformed plots. Therefore I was wondering whether only the intercepts need to be back-transformed, whereas the other effects are not affected by the transformation in first place (given that they are usually normally distributed anyway). Does that make sense?

![image](https://github.com/lnccbrown/HSSM/assets/8819465/bd1ca988-b9b4-4956-bbf8-a83696469f9a)

**Comment by eort** [comment]
Okay, I am more confused now. Looking at the same regression on the `g` parameter in the ornstein model, I get very weird (but reliable) estimates. Same regression means: `"g ~ C(drug, levels=lvl) + (1 | participant_id)"`

Here without transformation:
![image](https://github.com/lnccbrown/HSSM/assets/8819465/1cca8fbd-c3d4-47c4-a019-0200f776bb6f)

And here with:
![image](https://github.com/lnccbrown/HSSM/assets/8819465/adf6260e-a627-4697-8cc9-9b0db95a5de5)

For the intercept, it is clear that the transformation is necessary and the result reasonable, but for the drug effect the not-back-transformed data just don't make any sense (as it is beyond the bounds of `g`), and the back-transformed data appear to be somewhat unrealistic, even considering the now ignored random effects. Now I am wondering how to interpret the effects in regressions in first place.

**Comment by AlexanderFengler** [comment]
Hey @eort ,

I think the confusion is that you should basically run the transform on the output of the regression, not on single terms. Transforms operate as `f(X@b)` (where `X@b` is your regression equation). My original answer may have been slightly misleading on that front, but what I meant was to run `linkinv` for each parameter (`v`, `a`, `z` etc.), not each component of a regression.

If you look at the traces, you should see a trialwise `g`'s in there (something like `g[0], ..., g[N]`). Those are the ones you want to look at basically (these are post-transform --> back to original space), and those are the ones I suggested to manually compute.

However I will grant that those estimates of `g` are generally a bit weird. How do I interpret the scaling there? Without transformation, both `g_Intercept` and `g_C` are on the same scale roughly, but with transformation they are not?


Best,
Alex

**Reply by eort** [reply]
Thanks for the clarification!
To produce the plots  I used the `transform` keyword in the `arviz` plotting function. So for `a`, I did:

`az.plot_posterior(model._inference_obj, var_names='^a', filter_vars='regex', transform=model.link['a'].linkinv)`

I don't know whether arviz does it the correct way (`f (X@b)`) or not. 

>  How do I interpret the scaling there? Without transformation, both g_Intercept and g_C are on the same scale roughly, but with transformation they are not?

Not sure what you mean here. Without transformation the values are outside the domain in which that parameter is defined, no? Isn't it `delta_evidence = drift + g * current_evidence`? I have trouble wrapping my head around what values < -1 would mean here. With transformation the range is at least more reasonable. 

As far as I understood these regression situations, they don't necessarily need to be on the same scale. In this case, the intercept reflects `g` in case of the baseline condition of `drug`, whereas the drug effect is then the shift in the  `cyc` condition relative to baseline. So if interested in the "absolute" value of `g` in the `cyc` condition, I would add them up. Of course the fact that we are looking at distributions and not point estimates here prevents a straight forward summation of the.

**Comment by eort** [comment]
Hey, 

Sorry to bring this up again, but I still haven't managed to make sense of this. 

@AlexanderFengler, you mentioned that I should look at the trialwise parameters as they are post-transform. I managed and the distributions look alright, however, they incorporate of course all the fixed and random effects. It is unclear to me how I would extract the group effects of a condition from the trialwise estimates. If I don't care about the subject-specific effects or intercepts, how can I obtain estimates of the experimental factors in a space that interpretable for respective parameter?

Also, turning back to your point that I should run the transform on the output of the regression, not on single terms. If I understand you correctly, this would produce the trialwise estimates as in `g[0], g[1], ...`, so I would again end up in a situation where I cannot separate group from individual effects, wouldn't I?

This being said, I am increasingly sure that my implementations of the back-transformation (either via `arviz.map` or as argument in the plotting function), applies the transforms per term and not the full thing. Are these convenience functionalities that you mentioned already included somewhere in the code base?

Thanks again, 
Eduard

**Comment by eort** [comment]
Me again...

I looked up how Bambi is using link functions and came across this [example](https://bambinos.github.io/bambi/notebooks/beta_regression.html#beta-regression-with-predictors) where `logit` link functions are used (the link is slightly different, i.e. bounded by 0 and 1, but that shouldn't matter to illustrate the principle). Specifically, this part caught my attention:
![image](https://github.com/lnccbrown/HSSM/assets/8819465/4691e984-11be-4c3b-a5eb-93e57eabc446)

So, in this example the estimates are actually simply back-transformed and interpreted subsequently. Assuming that they know what they are doing, why is this permitted? Only reason I can think of is that in that example only a single regressor was put into the model, but than there is still the intercept. How does this example related to your (@AlexanderFengler) statement that applying the back-transform to individual terms does not make sense?

On an unrelated note, do I understand correctly that using a `logit` link function is essentially nothing else than turning my regression into a logistic regression (with a generalized link function)? If so, can't I just interpret the estimates similarly to log-odds?

Eduard
